insta/src/runtime.rs

```diff
@@ -644,50 +678,117 @@ pub fn assert_snapshot(
+                })
+                .unwrap_or(false);
+
+            if pass {
+                ctx.cleanup_passing()?;
+
+                if tool_config.force_update_snapshots() {
+                    ctx.update_snapshot(&mut new_snapshot)?;
+                }
+            // otherwise print information and update snapshots.
             } else {
-                x.matches(&new_snapshot)
+                print_snapshot_info(&ctx, &new_snapshot);
+                let update_result = ctx.update_snapshot(&mut new_snapshot)?;
+                finalize_assertion(&ctx, update_result);
             }
-        })
-        .unwrap_or(false);
+        }
+        SnapshotValue::Binary {
+            write: new_snapshot_value,
+            extension,
+        } => {
+            assert!(
+                !["new", "_"].contains(&extension),
+                "this file extension is not allowed",
+            );
+            assert!(
+                !extension.starts_with("new."),
+                "file extensions starting with 'new.' are not allowed",
+            );
+
+            let content = ctx.pre_create_binary_snapshot(new_snapshot_value, extension)?;
+
+            let new_binary_path = if let SnapshotContents::Binary { ref path, .. } = content {
+                path.clone()
+            } else {
+                unreachable!();
+            };
 
-    if pass {
-        ctx.cleanup_passing()?;
+            let mut new_snapshot = ctx.new_snapshot(content, expr);
 
-        if tool_config.force_update_snapshots() {
-            ctx.update_snapshot(new_snapshot)?;
+            // If we allow assertion with duplicates, we record the duplicate now.  This will
+            // in itself fail the assertion if the previous visit of the same assertion macro
+            // did not yield the same result.
+            RECORDED_DUPLICATES.with(|x| {
+                if let Some(results) = x.borrow_mut().last_mut() {
+                    record_snapshot_duplicate(results, &new_snapshot, &ctx);
+                }
+            });
+
+            let pass = ctx
+                .old_snapshot
+                .as_ref()
+                .map(|x| {
+                    if tool_config.require_full_match() {
+                        x.matches_fully(&new_snapshot)
+                    } else {
+                        x.matches(&new_snapshot)
+                    }
+                })
+                .unwrap_or(false);
+
+            if pass {
+                ctx.cleanup_passing()?;
+
+                if tool_config.force_update_snapshots() {
+                    ctx.update_snapshot(&mut new_snapshot)?;
+                }
+
+                fs::remove_file(new_binary_path)?;
+            // otherwise print information and update snapshots.
+            } else {
+                let update_result = ctx.update_snapshot(&mut new_snapshot)?;
+                print_snapshot_info(&ctx, &new_snapshot);
+                finalize_assertion(&ctx, update_result);
+            }
```

Comment:
Is it possible to reduce the amount of code duplication here?

(I wrote some of this though tbc am not a maintainer)

Reply:
Most of that was actually totally unnecessary.